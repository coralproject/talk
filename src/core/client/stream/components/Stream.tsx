import { Localized } from "fluent-react/compat";
import * as React from "react";
import { StatelessComponent } from "react";

import { Button, Flex } from "talk-ui/components";

import CommentContainer from "../containers/CommentContainer";
import PostCommentFormContainer from "../containers/PostCommentFormContainer";
import ReplyListContainer from "../containers/ReplyListContainer";
import UserBoxContainer from "../containers/UserBoxContainer";
import * as styles from "./Stream.css";

export interface StreamProps {
  assetID: string;
  isClosed?: boolean;
  comments: ReadonlyArray<{ id: string }>;
  onLoadMore?: () => void;
  hasMore?: boolean;
  disableLoadMore?: boolean;
}

const Stream: StatelessComponent<StreamProps> = props => {
  return (
    <Flex className={styles.root} direction="column" itemGutter>
      <UserBoxContainer />
      <PostCommentFormContainer assetID={props.assetID} />
      <Flex
        direction="column"
        id="talk-comments-stream-log"
        role="log"
        aria-live="polite"
        itemGutter
      >
        {props.comments.map(comment => (
          <Flex direction="column" key={comment.id} itemGutter>
            <CommentContainer data={comment} />
            <ReplyListContainer comment={comment} />
          </Flex>
        ))}
        {props.hasMore && (
          <Localized id="comments-stream-loadMore">
            <Button
              id={"talk-comments-stream-loadMore"}
              onClick={props.onLoadMore}
              variant="outlined"
              fullWidth
              disabled={props.disableLoadMore}
              aria-controls="talk-comments-stream-log"
            >
              Load More
            </Button>
          </Localized>
        )}
      </Flex>
    </Flex>
  );
};

export default Stream;
