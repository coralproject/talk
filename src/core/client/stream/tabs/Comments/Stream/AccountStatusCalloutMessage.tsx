import React, { FunctionComponent } from "react";

import styles from "./AccountStatusCalloutMessage.css";

interface Props {
  message: string;
}

const AccountStatusCalloutMessage: FunctionComponent<Props> = ({ message }) => {
  return <blockquote className={styles.message}>{message}</blockquote>;
};

export default AccountStatusCalloutMessage;
