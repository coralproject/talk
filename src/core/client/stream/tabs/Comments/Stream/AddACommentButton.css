.root {
  text-align: center;
}

.button {
  color: var(--palette-primary-500);
  margin: var(--spacing-4) auto;
}

.icon {
  margin-right: var(--spacing-1);
}
