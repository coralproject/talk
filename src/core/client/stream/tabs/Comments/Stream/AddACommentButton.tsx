import React, { FunctionComponent, useCallback } from "react";

import { useCoralContext } from "coral-framework/lib/bootstrap";
import { POST_COMMENT_FORM_ID } from "coral-stream/constants";
import { Flex, Icon } from "coral-ui/components/v2";
import { Button } from "coral-ui/components/v3";

import styles from "./AddACommentButton.css";

interface Props {
  isQA?: boolean;
}

const AddACommentButton: FunctionComponent<Props> = ({ isQA = false }) => {
  const { pym } = useCoralContext();
  const onClick = useCallback(() => {
    if (!pym) {
      return;
    }

    pym.scrollParentToChildEl(POST_COMMENT_FORM_ID);
  }, [pym]);

  return (
    <div className={styles.root}>
      <Button
        variant="outlined"
        color="primary"
        paddingSize="small"
        upperCase
        className={styles.button}
        onClick={onClick}
      >
        <Flex alignItems="center">
          <Icon className={styles.icon}>create</Icon>
          {isQA ? <span>Ask a Question</span> : <span>Add a Comment</span>}
        </Flex>
      </Button>
    </div>
  );
};

export default AddACommentButton;
