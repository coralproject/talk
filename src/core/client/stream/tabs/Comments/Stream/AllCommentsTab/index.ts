export { default, default as AllCommentsTabQuery } from "./AllCommentsTabQuery";
