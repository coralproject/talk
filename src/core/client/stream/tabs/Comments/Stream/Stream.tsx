import { Localized } from "fluent-react/compat";
import * as React from "react";
import { FunctionComponent } from "react";

import { PropTypesOf } from "coral-framework/types";
import UserBoxContainer from "coral-stream/common/UserBox";
import { Button, Flex, HorizontalGutter, Spinner } from "coral-ui/components";

import CommentContainer from "../Comment";
import ReplyListContainer from "../ReplyList";
import TombstoneOrHideContainer from "../TombstoneOrHideContainer";
import BannedInfo from "./BannedInfo";
import CommunityGuidelinesContainer from "./CommunityGuidelines";
import PostCommentFormContainer from "./PostCommentForm";
import SortMenu from "./SortMenu";

import styles from "./Stream.css";

export interface StreamProps {
  story: {
    id: string;
    isClosed?: boolean;
  } & PropTypesOf<typeof CommentContainer>["story"] &
    PropTypesOf<typeof ReplyListContainer>["story"] &
    PropTypesOf<typeof PostCommentFormContainer>["story"];
  settings: PropTypesOf<typeof CommentContainer>["settings"] &
    PropTypesOf<typeof ReplyListContainer>["settings"] &
    PropTypesOf<typeof UserBoxContainer>["settings"] &
    PropTypesOf<typeof CommunityGuidelinesContainer>["settings"] &
    PropTypesOf<typeof PostCommentFormContainer>["settings"] & {
      reaction: {
        sortLabel: string;
      };
    };
  comments: ReadonlyArray<
    { id: string } & PropTypesOf<typeof CommentContainer>["comment"] &
      PropTypesOf<typeof ReplyListContainer>["comment"] &
      PropTypesOf<typeof TombstoneOrHideContainer>["comment"]
  >;
  onLoadMore?: () => void;
  hasMore?: boolean;
  disableLoadMore?: boolean;
  viewer:
    | PropTypesOf<typeof UserBoxContainer>["viewer"] &
        PropTypesOf<typeof CommentContainer>["viewer"] &
        PropTypesOf<typeof ReplyListContainer>["viewer"] &
        PropTypesOf<typeof TombstoneOrHideContainer>["viewer"]
    | null;
  orderBy: PropTypesOf<typeof SortMenu>["orderBy"];
  onChangeOrderBy: (e: React.ChangeEvent<HTMLSelectElement>) => void;
  refetching?: boolean;
  banned?: boolean;
}

const Stream: FunctionComponent<StreamProps> = props => {
  return (
    <HorizontalGutter className={styles.root} size="double">
      <UserBoxContainer viewer={props.viewer} settings={props.settings} />
      <CommunityGuidelinesContainer settings={props.settings} />
      {!props.banned && (
        <PostCommentFormContainer
          settings={props.settings}
          story={props.story}
        />
      )}
      {props.banned && <BannedInfo />}
      {props.comments.length > 0 && (
        <SortMenu
          orderBy={props.orderBy}
          onChange={props.onChangeOrderBy}
          reactionSortLabel={props.settings.reaction.sortLabel}
        />
      )}
      {props.refetching && (
        <Flex justifyContent="center">
          <Spinner />
        </Flex>
      )}
      {!props.refetching && (
        <HorizontalGutter
          id="coral-comments-stream-log"
          data-testid="comments-stream-log"
          role="log"
          aria-live="polite"
        >
          {props.comments.map(comment => (
            <TombstoneOrHideContainer
              key={comment.id}
              viewer={props.viewer}
              comment={comment}
            >
              <HorizontalGutter>
                <CommentContainer
                  viewer={props.viewer}
                  settings={props.settings}
                  comment={comment}
                  story={props.story}
                />
                <ReplyListContainer
                  settings={props.settings}
                  viewer={props.viewer}
                  comment={comment}
                  story={props.story}
                />
              </HorizontalGutter>
            </TombstoneOrHideContainer>
          ))}
          {props.hasMore && (
            <Localized id="comments-stream-loadMore">
              <Button
                id={"coral-comments-stream-loadMore"}
                onClick={props.onLoadMore}
                variant="outlined"
                fullWidth
                disabled={props.disableLoadMore}
                aria-controls="coral-comments-stream-log"
              >
                Load More
              </Button>
            </Localized>
          )}
        </HorizontalGutter>
      )}
    </HorizontalGutter>
  );
};

export default Stream;
