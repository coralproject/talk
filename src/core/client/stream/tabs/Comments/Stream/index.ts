export { default, default as StreamQuery } from "./StreamQuery";
