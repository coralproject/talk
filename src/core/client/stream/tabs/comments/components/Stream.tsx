import { Localized } from "fluent-react/compat";
import * as React from "react";
import { StatelessComponent } from "react";

import { PropTypesOf } from "talk-framework/types";
import UserBoxContainer from "talk-stream/containers/UserBoxContainer";
import { Button, HorizontalGutter } from "talk-ui/components";

import CommentContainer from "../containers/CommentContainer";
import PostCommentFormContainer from "../containers/PostCommentFormContainer";
import ReplyListContainer from "../containers/ReplyListContainer";
import PostCommentFormFake from "./PostCommentFormFake";
import * as styles from "./Stream.css";

export interface StreamProps {
  asset: {
    id: string;
    isClosed?: boolean;
  } & PropTypesOf<typeof CommentContainer>["asset"] &
    PropTypesOf<typeof ReplyListContainer>["asset"];
  settings: PropTypesOf<typeof CommentContainer>["settings"] &
    PropTypesOf<typeof ReplyListContainer>["settings"];
  comments: ReadonlyArray<
    { id: string } & PropTypesOf<typeof CommentContainer>["comment"] &
      PropTypesOf<typeof ReplyListContainer>["comment"]
  >;
  onLoadMore?: () => void;
  hasMore?: boolean;
  disableLoadMore?: boolean;
  me:
    | PropTypesOf<typeof UserBoxContainer>["me"] &
        PropTypesOf<typeof CommentContainer>["me"] &
        PropTypesOf<typeof ReplyListContainer>["me"]
    | null;
}

const Stream: StatelessComponent<StreamProps> = props => {
  return (
    <HorizontalGutter className={styles.root} size="double">
      <HorizontalGutter size="half">
        <UserBoxContainer me={props.me} />
        {props.me ? (
          <PostCommentFormContainer assetID={props.asset.id} />
        ) : (
          <PostCommentFormFake />
        )}
      </HorizontalGutter>
      <HorizontalGutter
        id="talk-comments-stream-log"
        role="log"
        aria-live="polite"
      >
        {props.comments.map(comment => (
          <HorizontalGutter key={comment.id}>
            <CommentContainer
              me={props.me}
              settings={props.settings}
              comment={comment}
              asset={props.asset}
            />
            <ReplyListContainer
              settings={props.settings}
              me={props.me}
              comment={comment}
              asset={props.asset}
            />
          </HorizontalGutter>
        ))}
        {props.hasMore && (
          <Localized id="comments-stream-loadMore">
            <Button
              id={"talk-comments-stream-loadMore"}
              onClick={props.onLoadMore}
              variant="outlined"
              fullWidth
              disabled={props.disableLoadMore}
              aria-controls="talk-comments-stream-log"
            >
              Load More
            </Button>
          </Localized>
        )}
      </HorizontalGutter>
    </HorizontalGutter>
  );
};

export default Stream;
