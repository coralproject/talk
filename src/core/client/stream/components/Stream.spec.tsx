import { shallow } from "enzyme";
import { noop } from "lodash";
import React from "react";
import sinon, { SinonSpy } from "sinon";

import { removeFragmentRefs } from "talk-framework/testHelpers";
import { PropTypesOf } from "talk-framework/types";

import Stream from "./Stream";

const StreamN = removeFragmentRefs(Stream);

it("renders correctly", () => {
  const props: PropTypesOf<typeof StreamN> = {
    assetID: "asset-id",
    isClosed: false,
    comments: [{ id: "comment-1" }, { id: "comment-2" }],
    onLoadMore: noop,
    disableLoadMore: false,
    hasMore: false,
    user: null,
  };
  const wrapper = shallow(<StreamN {...props} />);
  expect(wrapper).toMatchSnapshot();
});

describe("when use is logged in", () => {
  it("renders correctly", () => {
    const props: PropTypesOf<typeof StreamN> = {
      assetID: "asset-id",
      isClosed: false,
      comments: [{ id: "comment-1" }, { id: "comment-2" }],
      onLoadMore: noop,
      disableLoadMore: false,
      hasMore: false,
      user: {},
    };
    const wrapper = shallow(<StreamN {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("when there is more", () => {
  const props: PropTypesOf<typeof StreamN> = {
    assetID: "asset-id",
    isClosed: false,
    comments: [{ id: "comment-1" }, { id: "comment-2" }],
    onLoadMore: sinon.spy(),
    disableLoadMore: false,
    hasMore: true,
    user: null,
  };

  const wrapper = shallow(<StreamN {...props} />);
  it("renders a load more button", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("calls onLoadMore", () => {
    wrapper.find("#talk-comments-stream-loadMore").simulate("click");
    expect((props.onLoadMore as SinonSpy).calledOnce).toBe(true);
  });

  const wrapperDisabledButton = shallow(<StreamN {...props} disableLoadMore />);
  it("disables load more button", () => {
    expect(wrapperDisabledButton).toMatchSnapshot();
  });
});
