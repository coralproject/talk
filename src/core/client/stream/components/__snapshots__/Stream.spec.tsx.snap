// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Stream-root"
>
  <Logo
    gutterBottom={true}
  />
  <withContext(createMutationContainer(PostCommentFormContainer))
    assetID="asset-id"
  />
  <withPropsOnChange(Flex)
    aria-live="polite"
    direction="column"
    id="talk-comments-stream-log"
    itemGutter={true}
    role="log"
  >
    <withPropsOnChange(Flex)
      direction="column"
      itemGutter={true}
      key="comment-1"
    >
      <Relay(CommentContainer)
        data={
          Object {
            "id": "comment-1",
          }
        }
      />
      <Relay(ReplyListContainer)
        comment={
          Object {
            "id": "comment-1",
          }
        }
      />
    </withPropsOnChange(Flex)>
    <withPropsOnChange(Flex)
      direction="column"
      itemGutter={true}
      key="comment-2"
    >
      <Relay(CommentContainer)
        data={
          Object {
            "id": "comment-2",
          }
        }
      />
      <Relay(ReplyListContainer)
        comment={
          Object {
            "id": "comment-2",
          }
        }
      />
    </withPropsOnChange(Flex)>
  </withPropsOnChange(Flex)>
</div>
`;

exports[`when there is more disables load more button 1`] = `
<div
  className="Stream-root"
>
  <Logo
    gutterBottom={true}
  />
  <withContext(createMutationContainer(PostCommentFormContainer))
    assetID="asset-id"
  />
  <withPropsOnChange(Flex)
    aria-live="polite"
    direction="column"
    id="talk-comments-stream-log"
    itemGutter={true}
    role="log"
  >
    <withPropsOnChange(Flex)
      direction="column"
      itemGutter={true}
      key="comment-1"
    >
      <Relay(CommentContainer)
        data={
          Object {
            "id": "comment-1",
          }
        }
      />
      <Relay(ReplyListContainer)
        comment={
          Object {
            "id": "comment-1",
          }
        }
      />
    </withPropsOnChange(Flex)>
    <withPropsOnChange(Flex)
      direction="column"
      itemGutter={true}
      key="comment-2"
    >
      <Relay(CommentContainer)
        data={
          Object {
            "id": "comment-2",
          }
        }
      />
      <Relay(ReplyListContainer)
        comment={
          Object {
            "id": "comment-2",
          }
        }
      />
    </withPropsOnChange(Flex)>
    <Localized
      id="comments-stream-loadMore"
    >
      <withPropsOnChange(Button)
        aria-controls="talk-comments-stream-log"
        disabled={true}
        fullWidth={true}
        id="talk-comments-stream-loadMore"
        onClick={[Function]}
        variant="outlined"
      >
        Load More
      </withPropsOnChange(Button)>
    </Localized>
  </withPropsOnChange(Flex)>
</div>
`;

exports[`when there is more renders a load more button 1`] = `
<div
  className="Stream-root"
>
  <Logo
    gutterBottom={true}
  />
  <withContext(createMutationContainer(PostCommentFormContainer))
    assetID="asset-id"
  />
  <withPropsOnChange(Flex)
    aria-live="polite"
    direction="column"
    id="talk-comments-stream-log"
    itemGutter={true}
    role="log"
  >
    <withPropsOnChange(Flex)
      direction="column"
      itemGutter={true}
      key="comment-1"
    >
      <Relay(CommentContainer)
        data={
          Object {
            "id": "comment-1",
          }
        }
      />
      <Relay(ReplyListContainer)
        comment={
          Object {
            "id": "comment-1",
          }
        }
      />
    </withPropsOnChange(Flex)>
    <withPropsOnChange(Flex)
      direction="column"
      itemGutter={true}
      key="comment-2"
    >
      <Relay(CommentContainer)
        data={
          Object {
            "id": "comment-2",
          }
        }
      />
      <Relay(ReplyListContainer)
        comment={
          Object {
            "id": "comment-2",
          }
        }
      />
    </withPropsOnChange(Flex)>
    <Localized
      id="comments-stream-loadMore"
    >
      <withPropsOnChange(Button)
        aria-controls="talk-comments-stream-log"
        disabled={false}
        fullWidth={true}
        id="talk-comments-stream-loadMore"
        onClick={[Function]}
        variant="outlined"
      >
        Load More
      </withPropsOnChange(Button)>
    </Localized>
  </withPropsOnChange(Flex)>
</div>
`;
